File: backend/models/Product.js
const mongoose = require('mongoose');
const ProductSchema = new mongoose.Schema({
    barcode: { type: String, required: true, unique: true },
    name: { type: String, required: true },
    category: { type: String, required: true, enum: ['plastic', 'paper', 'tin', 'glass', 'other'] },
    manufacturer: { type: String, required: true },
    manufactureDate: { type: Date, required: true },
    distributionHistory: [{ location: String, date: Date }],
    currentStatus: { type: String, default: 'manufactured', enum: ['manufactured', 'distributed', 'sold', 'in_use', 'ready_for_disposal', 'disposed'] },
    owner: { type: mongoose.Schema.Types.ObjectId, ref: 'User' }, // Consumer who scanned it
    usageNotes: { type: String },
    blockchainHistory: [{ // Simplified reference to blockchain transactions
        txHash: String,
        event: String,
        timestamp: Date
    }]
}, { timestamps: true });
module.exports = mongoose.model('Product', ProductSchema);

File: flutter_frontend/lib/models/product.dart
class Product {
  final String barcode;
  final String name;
  final String category;
  final String manufacturer;
  final String manufactureDate;
  final String? owner;
  final String currentStatus;
  final List<Map<String, dynamic>>? distributionHistory;

  Product({
    required this.barcode,
    required this.name,
    required this.category,
    required this.manufacturer,
    required this.manufactureDate,
    this.owner,
    required this.currentStatus,
    this.distributionHistory,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      barcode: json['barcode'] ?? '',
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      manufacturer: json['manufacturer'] ?? '',
      manufactureDate: json['manufactureDate'] ?? '',
      owner: json['owner'],
      currentStatus: json['currentStatus'] ?? 'unknown',
      distributionHistory: json['distributionHistory'] != null
          ? List<Map<String, dynamic>>.from(json['distributionHistory'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'barcode': barcode,
      'name': name,
      'category': category,
      'manufacturer': manufacturer,
      'manufactureDate': manufactureDate,
      'owner': owner,
      'currentStatus': currentStatus,
      'distributionHistory': distributionHistory,
    };
  }
}
